File: win/src/LinkOpener.cs
using System;

namespace Sanity
{
    public static class LinkOpener
    {
        public static bool Open(string url)
        {
            if (string.IsNullOrWhiteSpace(url))
                return false;

            var config = AppConfig.Load();
            var finalUrl = url.Trim();
            var cleaned = false;

            if (config.IsLinkProxyActive)
            {
                string cleanedUrl;
                if (UrlCleaner.TryClean(finalUrl, config.Rules, out cleanedUrl))
                {
                    finalUrl = cleanedUrl;
                    cleaned = true;
                    UsageMetrics.RecordClean(finalUrl);
                }
            }

            var targetBrowser = ResolveTargetBrowser(config);
            BrowserHelper.TryLaunchBrowser(targetBrowser, finalUrl);
            return cleaned;
        }

        private static string ResolveTargetBrowser(AppConfig config)
        {
            if (!string.IsNullOrWhiteSpace(config.TargetBrowser))
                return config.TargetBrowser;

            return ProtocolRegistration.GetBackedUpBrowser() ?? string.Empty;
        }
    }
}

File: mac/Sanity/Sources/LinkOpener.swift
import Foundation

enum LinkOpener {
    @discardableResult
    static func open(_ urlString: String) -> Bool {
        let config = AppConfig.load()
        var finalUrl = urlString.trimmingCharacters(in: .whitespacesAndNewlines)
        var cleaned = false

        if config.isLinkProxyActive,
           let cleanedUrl = UrlCleaner.tryClean(finalUrl, rules: config.rules) {
            finalUrl = cleanedUrl
            cleaned = true
            UsageMetrics.recordClean(url: finalUrl)
        }

        BrowserHelper.open(url: finalUrl, targetBrowser: resolveTargetBrowser(config: config))
        UpdateChecker.runSync(config: config)
        return cleaned
    }

    private static func resolveTargetBrowser(config: AppConfig) -> String {
        if !config.targetBrowser.isEmpty {
            return config.targetBrowser
        }
        return UserDefaults.standard.string(forKey: "previousHttpHandler") ?? ""
    }
}
